packages/opentelemetry-resources/src/Resource.ts

```diff
@@ -54,8 +58,50 @@ export class Resource {
      * information about the entity as numbers, strings or booleans
      * TODO: Consider to add check/validation on attributes.
      */
-    readonly attributes: ResourceAttributes
-  ) {}
+    attributes: ResourceAttributes,
+    asyncAttributesPromise?: Promise<ResourceAttributes>
+  ) {
+    this._attributes = attributes;
+    this._asyncAttributesHaveResolved = asyncAttributesPromise === undefined;
+    this._asyncAttributesPromise = asyncAttributesPromise?.then(
+      asyncAttributes => {
+        this._attributes = Object.assign({}, this._attributes, asyncAttributes);
+        this._asyncAttributesHaveResolved = true;
+        return asyncAttributes;
+      }
+    );
+    this._asyncAttributesPromise?.catch(err => {
```

Comment:
This `catch` call is creating a new promise chain. Since the catch handler is returning an empty resource, we can combine the `then` call with the catch handler. In this way, the logging at the call site of `waitForAsyncAttributes` in sdk-trace and sdk-metrics can also be removed.

```suggestion
    this._asyncAttributesPromise = asyncAttributesPromise?.then(
      asyncAttributes => {
        this._attributes = Object.assign({}, this._attributes, asyncAttributes);
        this._asyncAttributesHaveResolved = true;
        return asyncAttributes;
      },
      err => {
        diag.debug("The resource's async promise rejected: %s", err);
        this._asyncAttributesHaveResolved = true;
        return {};
      }
    );
```

This should also align with the current `detectResource` behavior: return an empty resource if there is any exceptions and not propagating the error to the caller.